### PR TITLE
Make `property.Value` immutable

### PR DESCRIPTION
**Pull request by iwahbe**
tldr: making all `property.Value`s immutable will make the API less idiomatic, but it will make API design and usage of `property.Value` clearer and less bug prone. We can do it without sacrificing too much performance.

This change is stacked on top of https://github.com/pulumi/pulumi/pull/17374.

There are two major concerns to address with this PR:

- The API is unidiomatic. ☝️ explains why I think the change is worth it from that perspective.
- The change is not performant.

## API Usage

Regardless of this PR, all APIs that manipulate `property.Value` *must* be of the shape:

```go
func alterValue(startingValue property.Value) (mutatedValue property.Value, _ error)
```

This is because you cannot reliably mutate `startingValue`. You can only mutate `startingValue` if it is a map or an array.[^1] APIs of this shape essentially must discard `startingValue` after `alterValue` is called, since they cannot be sure what value it holds, or what relation it has with `mutatedValue`.

Other languages can express this as an inout (C++) value, or as passing an owned value in and requiring an owned value out (Rust). In Go, we could express this with an in-out `*property.Value`, but that is highly unidiomatic:

```go
func alterValue(value *property.Value) error
```

By making `property.Value` immutable, we side-step this entire conversation. The idiomatic Go function is correct:

```go
func alterValue(property.Value) (property.Value, _ error)
```

The calling API is now free to reuse the value it passed in. The immutable nature of the API makes APIs like `func(property.Value)` clearly wrong, as opposed to flakey.

## Performance

As long as we keep all values immutable, we only need shallow copies on `.AsMap/.AsArray`, and *never* need to make deep copies. `property.Value.Equals` can depend on internals to make target deep mutations as efficient as possible. A full manual traversal of a nested datastructure has the same cost as copying.

I believe that we can afford the additional cost of copying to prevent bugs. TF's underlying data value is immutable (implemented the same way) and TF does not suffer on performance. IMO, the immutability of `cty.Value` makes it much easier to work with.

[^1]: If you're sure that the func will only accept an array or a map, it should encode that in the type system.

### CODE REVIEW

**Comment by iwahbe** [comment]
This is a breaking change, but given existing use of `property.Value`, I think it is safe.

**Comment by tgummerer** [comment]
> This is because you cannot reliably mutate `startingValue`. You can only mutate `startingValue` if it is a map or an array.

Maybe I'm missing something here, but why is that the case?  `property.Value` is a struct, so we can modify it just fine, as long as we have the pointer to it.  

```
func alterValue(value *property.Value) error
```

I'm not sure I follow why :point_up: is unidiomatic?  That seems like the right way to do this in Go to me.  If you just have a `property.Value` you want to change it might even be 

```
func (v *property.Value) Set(val any) error
```
?

**Comment by iwahbe** [comment]
> Maybe I'm missing something here, but why is that the case? property.Value is a struct, so we can modify it just fine, as long as we have the pointer to it.

As long as we have a pointer to it, yes. I'm saying that `propert.Value` it itself only mutable if it contains a `property.Map` or `property.Array` as the inner type.

> I'm not sure I follow why ☝️ is unidiomatic? That seems like the right way to do this in Go to me. If you just have a property.Value you want to change it might even be
>
>```go
>func (v *property.Value) Set(val any) error
>```

That's now what happens in practice. We see plenty of APIs designed as in & out (instead of inout) vars:

```
pkg/engine/events.go
657:func filterPropertyValue(v resource.PropertyValue, debug bool) resource.PropertyValue {
128:func pvApply(pv resource.PropertyValue, f func(resource.PropertyValue) resource.PropertyValue) resource.PropertyValue {
pkg/resource/stack/deployment_test.go
628:func replaceOutputsWithComputed(v resource.PropertyValue) resource.PropertyValue {
pkg/resource/deploy/source_eval.go
2640:   var downgradeOutputPropertyValue func(v resource.PropertyValue) resource.PropertyValue
2642:   downgradeOutputPropertyValue = func(v resource.PropertyValue) resource.PropertyValue {
pkg/backend/display/json.go
37:func massagePropertyValue(v resource.PropertyValue, showSecrets bool) resource.PropertyValue {
```

These were examples I pulled from pulumi/pulumi. I see no examples of `*resource.PropertyValue` used as function inputs in pulumi/pulumi.

The bridge has the same problem. Given that in & out variables are how we actually use property values, I think immutability would help enforce better API design (also, it provides all the normal immutability benefits).

File: sdk/go/property/array.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package property
+
+// An immutable Array of [Value]s.
+type Array struct{ arr []Value }
```
Comment (Frassle) [inline sdk/go/property/array.go:18]: Worth adding Append: Array -> Value -> Array?

Comment (iwahbe) [reply sdk/go/property/array.go:18]: Sure.

**Review by Frassle — APPROVED** [review]
